build-package/certifaiReferenceModelServer/utils/local_server.py

```diff
@@ -7,25 +7,74 @@
 import traceback
 import importlib
 from inspect import getmembers, isfunction
-from flask import Flask, Response, request, jsonify
+from typing import NamedTuple, Dict, List
+
+from flask import Flask, Response, request
 from cortex import Message
+import json
+from werkzeug.exceptions import BadRequest
 
 app = Flask(__name__)
 
 
+class EmptyResponse(NamedTuple):
+    response: Dict = {"predictions": []}
+
+
+class EmptyRequest(NamedTuple):
+    request: Dict = {"payload": {"instances": []}}
+
+
+def is_empty(a):
+    if isinstance(a, List):
+        return all(map(is_empty, a))
+    return False
+
+
+def _validate_message(msg):
+    try:
+        if msg.payload is None:
+            return BadRequest(f'Missing `payload`. Expected format {json.dumps(EmptyRequest().request)}')
+        elif not isinstance(msg.payload, Dict):
+            return BadRequest(
+                f'`payload` must be of type dict. Expected format {json.dumps(EmptyRequest().request)}')
+        elif 'instances' not in msg.payload.keys():
+            return BadRequest(f'Missing `instances`. Expected format {json.dumps(EmptyRequest().request)}')
+        elif not isinstance(msg.payload.get('instances'), List):
+            return BadRequest("`instances` must be of type List[Union[a, List a] ")
+        elif is_empty(msg.payload.get('instances')):
```

Comment:
Out of curiosity, Is it necessary to handle the empty case here? I would assume the model's `predict` function would handle it just fine. (If it does, then what do you think about removing this special case so the `_validate_message()` functions always returns an error/response of the same Type - coincidentally a `BadRequest` object.)

Reply:
Actually... would be better if it doesnt return an error in the empty list case. For at least a while, the scanner used to occasionally send empty predict requests. I know it's unlikely to be intended but I'd rather let it through....

Reply:
I thought about this for some time, predict doesn't handle the empty cases correctly (mostly due to old encoders and reshape, can be fixed easily but would have to change many predict.py files) and throws errors which are caught correctly(now) and returns error but ideally its not an error just empty response and so best case scenario is handle the case before using cpu cycles.